scripts/train_uqno_darcy.py

```diff
@@ -0,0 +1,585 @@
+    data_processor = data_processor.to(device)
+    for idx, sample in enumerate(loader):
+        sample = data_processor.preprocess(sample)
+        out = model(**sample)
+        out, sample = data_processor.postprocess(out, sample) # unnormalize output
+
+        x_list.append(sample['x'].to("cpu"))
+        error = (out-sample['y']).detach().to("cpu") # detach, otherwise residual carries gradient of model weight
+        # error is unnormalized here
+        error_list.append(error)
+        
+        del sample, out
+    errors = torch.cat(error_list, axis=0)
+    xs = torch.cat(x_list, axis=0) # check this
+    
+    residual_encoder = UnitGaussianNormalizer()
+    residual_encoder.fit(errors)
+    
+    # positional encoding and normalization already applied to X values
+    residual_data_processor = DefaultDataProcessor(in_normalizer=None,
+                                                   out_normalizer=residual_encoder)
+    residual_data_processor.train()
+
+    if train_val_split:
+        val_start = int(0.8 * xs.shape[0])
+
+        residual_train_db = TensorDataset(x=xs[:val_start], y=errors[:val_start])
+        residual_val_db = TensorDataset(x=xs[val_start:], y=errors[val_start:])
+    else:
+        residual_val_db = None
+    return residual_train_db, residual_val_db, residual_data_processor
+
+
+class UQNODataProcessor(DataProcessor):
+    def __init__(self, base_data_processor: DataProcessor, resid_data_processor: DataProcessor,
+                 device: str="cpu"):
+        """UQNODataProcessor converts tuple (G_hat(a,x), E(a,x)) and 
+        sample['y'] = G_true(a,x) into the form expected by PointwiseQuantileLoss
+
+        y_pred = E(a,x)
+        y_true = abs(G_hat(a,x) - G_true(a,x))
+
+        It also preserves any transformations that need to be performed
+        on inputs/outputs from the solution model. 
+
+        Parameters
+        ----------
+        base_data_processor : DataProcessor
+            transforms required for base solution_model input/output
+        resid_data_processor : DataProcessor
+            transforms required for residual input/output
+        device: str
+            "cpu" or "cuda" 
+        """
+        super().__init__()
+        self.base_data_processor = base_data_processor
+        self.residual_normalizer = resid_data_processor.out_normalizer
+
+        self.device = device
+        self.scale_factor = None
+    
+    def set_scale_factor(self, factor):
+        self.scale_factor = factor.to(device)
+    
+    def wrap(self, model):
+        self.model = model
+        return self
+
+    def to(self, device):
+        self.device = device
+        self.base_data_processor = self.base_data_processor.to(device)
+        self.residual_normalizer = self.residual_normalizer.to(device)
+        return self
+    
+    def train(self):
+        self.base_data_processor.train()
+    
+    def eval(self):
+        self.base_data_processor.eval()
```

Comment:
Same comment as above - let's check if this is needed.

Reply:
added a test